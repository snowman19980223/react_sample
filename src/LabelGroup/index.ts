export {default, LabelGroupProps} from './LabelGroup'
