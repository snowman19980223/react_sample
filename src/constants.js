import * as styledSystem from 'styled-system'
import systemPropTypes from '@styled-system/prop-types'
import themeGet from '@styled-system/theme-get'
import theme from './theme'

const {get: getKey, compose} = styledSystem

export const get = key => themeGet(key, getKey(theme, key))

export const COMMON = compose(
  styledSystem.space,
  styledSystem.color,
  styledSystem.display
)
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.color
}
export const BORDER = compose(
  styledSystem.border,
  styledSystem.shadow
)
BORDER.propTypes = {
  ...systemPropTypes.border,
  ...systemPropTypes.shadow
}

// these are 1:1 with styled-system's API now,
// so you could consider dropping the abstraction
export const TYPOGRAPHY = styledSystem.typography
export const LAYOUT = styledSystem.layout
export const POSITION = styledSystem.position
export const FLEX = styledSystem.flexbox

TYPOGRAPHY.propTypes = systemPropTypes.typography
LAYOUT.propTypes = systemPropTypes.layout
POSITION.propTypes = systemPropTypes.position
FLEX.propTypes = systemPropTypes.flexbox
