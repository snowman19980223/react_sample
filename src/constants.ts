import systemPropTypes from '@styled-system/prop-types'
import {themeGet} from '@styled-system/theme-get'
import PropTypes from 'prop-types'
import React from 'react'
import * as styledSystem from 'styled-system'
import theme from './theme'

interface StyleFn extends styledSystem.styleFn {
  propTypes?: React.WeakValidationMap<any>
}

const {get: getKey, compose, system} = styledSystem

export const get = (key: string) => themeGet(key, getKey(theme, key))

// Common props

export const COMMON: StyleFn = compose(styledSystem.space, styledSystem.color, styledSystem.display)

export interface SystemCommonProps
  extends styledSystem.ColorProps,
    styledSystem.SpaceProps,
    styledSystem.DisplayProps {}

COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.color
}

// Typography props

const whiteSpace = system({
  whiteSpace: {
    property: 'whiteSpace'
    // cssProperty: 'whiteSpace',
  }
})

export const TYPOGRAPHY: StyleFn = compose(styledSystem.typography, whiteSpace)

export interface SystemTypographyProps extends styledSystem.TypographyProps {
  whiteSpace?: 'normal' | 'nowrap' | 'pre' | 'pre-wrap' | 'pre-line'
}

TYPOGRAPHY.propTypes = {
  ...systemPropTypes.typography,
  whiteSpace: PropTypes.oneOf(['normal', 'nowrap', 'pre-wrap', 'pre', 'pre-line'])
}

// Border props

export const BORDER: StyleFn = compose(styledSystem.border, styledSystem.shadow)

export interface SystemBorderProps extends styledSystem.BorderProps, styledSystem.ShadowProps {}

BORDER.propTypes = {
  ...systemPropTypes.border,
  ...systemPropTypes.shadow
}

// Layout props

export const LAYOUT: StyleFn = styledSystem.layout

export type SystemLayoutProps = styledSystem.LayoutProps

LAYOUT.propTypes = systemPropTypes.layout

// Position props

export const POSITION: StyleFn = styledSystem.position

export type SystemPositionProps = styledSystem.PositionProps

POSITION.propTypes = systemPropTypes.position

// Flex props

export const FLEX: StyleFn = styledSystem.flexbox

export type SystemFlexProps = styledSystem.FlexboxProps

FLEX.propTypes = systemPropTypes.flexbox

// Grid props

export const GRID: StyleFn = styledSystem.grid

export type SystemGridProps = styledSystem.GridProps

GRID.propTypes = systemPropTypes.grid
