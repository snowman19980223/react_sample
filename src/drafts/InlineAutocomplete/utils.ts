import getCaretCoordinates from '@koddsson/textarea-caret'
import {Children, EventHandler, SyntheticEvent} from 'react'

import {Coordinates, ShowSuggestionsEvent, Suggestion, TextInputCompatibleChild, Trigger} from './types'

const singleWordTriggerTerminators = new Set([' ', '\n'])
const multiWordTriggerTerminators = new Set(['.', '\n'])

const isWhitespace = (char: string) => /\s/.test(char)

/**
 * Calculate whether or not suggestions should be shown based on the given state of the
 * input. If they should be shown, returns the show event.
 */
export const calculateSuggestionsQuery = (
  triggers: Array<Trigger>,
  text: string,
  caretLocation: number
): ShowSuggestionsEvent | null => {
  // Build backwards from the caret location until the most recent trigger character or terminator
  for (
    let i = caretLocation - 1, query = '', potentialTriggers = triggers;
    i >= 0 && potentialTriggers.length > 0;
    i--
  ) {
    const character = text[i]

    if (singleWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => t.multiWord)
    if (multiWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => !t.multiWord)

    for (const trigger of potentialTriggers.filter(t => character === t.triggerChar)) {
      // Trigger chars must always be preceded by whitespace or be the first character in the input,
      // and even a multi-word query cannot start with whitespace
      if ((i === 0 || isWhitespace(text[i - 1])) && !isWhitespace(query[0])) return {trigger, query}

      potentialTriggers = potentialTriggers.filter(t => t !== trigger)
    }

    query = character + query
  }

  return null
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}

export const getSuggestionValue = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.value

export const getSuggestionKey = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.key ?? suggestion.value

/**
 * Replace a section of a string.
 */
export const replaceSlice = (
  original: string,
  [startInclusive, endExclusive]: [number, number],
  replacement: string
) => {
  const before = original.substring(0, startInclusive)
  const after = original.substring(endExclusive)
  return before + replacement + after
}

/**
 * Attempts to assert that the child element is of a supported type. This can't be enforced
 * by the type system so it has to be done as a runtime check. This isn't foolproof - a
 * component that forwards a ref to a correct element but does not forward event handlers
 * will not work. But it's the best we can reasonably do.
 */
export function requireChildrenToBeInput(
  child: React.ReactElement,
  childRef: React.RefObject<HTMLElement>
): TextInputCompatibleChild {
  Children.only(child) // Assert that the child is lonely
  if (
    // There is no way to know what type the underlying child is until it mounts, so this
    // will always pass on first render before failing on the second render
    childRef.current &&
    !(childRef.current instanceof HTMLInputElement) &&
    !(childRef.current instanceof HTMLTextAreaElement)
  ) {
    throw new TypeError(
      `AutocompleteTextarea child must be a component that forwards a ref and props to an <input> or <textarea> element.`
    )
  }
  return child
}

/**
 * Combine several event handlers into one. The last handler in the list is called first
 * and no further handlers will be called if `event.preventDefault()` is called.
 */
export const augmentHandler =
  <E extends SyntheticEvent>(...handlers: Array<EventHandler<E> | undefined>) =>
  (event: E) => {
    for (const handler of [...handlers].reverse()) {
      if (!event.isDefaultPrevented()) handler?.(event)
    }
  }
