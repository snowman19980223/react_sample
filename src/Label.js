import PropTypes from 'prop-types'
import styled, {css} from 'styled-components'
import {variant} from 'styled-system'
import theme from './theme'
import {COMMON, get} from './constants'

const outlineStyles = css`
  margin-top: -1px; // offsets the 1px border
  margin-bottom: -1px; // offsets the 1px border
  font-weight: 400;
  color: ${get('colors.gray.6')};
  background-color: transparent;
  border: ${get('borders.1')} ${get('colors.blackfade15')};
  box-shadow: none;
`

const sizeVariant = variant({
  variants: {
    small: {
      fontSize: 0,
      px: '0.125em',
      py: 1
    },
    medium: {
      fontSize: 0,
      px: '3px',
      py: 1
    },
    large: {
      fontSize: 1,
      px: 1,
      py: 2
    },
    xl: {
      fontSize: 2,
      px: 1,
      py: 2
    }
  }
})

const Label = styled('span')`
  display: inline-block;
  font-weight: 600;
  line-height: ${get('lineHeights.condensedUltra')};
  color: ${get('colors.white')};
  border-radius: 2px;
  &:hover {
    text-decoration: none;
  }
  ${sizeVariant}
  ${COMMON} ${props => (props.dropshadow ? 'box-shadow: inset 0 -1px 0 rgba(27, 31, 35, 0.12)' : '')};
  ${props => (props.outline ? outlineStyles : '')}; // must be last to override other values
`

Label.defaultProps = {
  theme,
  bg: 'gray.5',
  variant: 'medium'
}

Label.propTypes = {
  dropshadow: PropTypes.bool,
  outline: PropTypes.bool,
  theme: PropTypes.object,
  variant: PropTypes.oneOf(['small', 'medium', 'large', 'xl']),
  ...COMMON.propTypes
}

export default Label
