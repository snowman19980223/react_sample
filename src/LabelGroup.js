import React from 'react'
import styled from 'styled-components'
import theme from './theme'
import {COMMON, get} from './constants'

const transformChildren = children => {
  return React.Children.map(children, child => {
    return React.cloneElement(child, {className: 'LabelItem'})
  })
}

const StyledLabelGroup = styled.span`
  ${COMMON}
  & .LabelItem {
    margin-right: ${get('space.1')}px;
  }
  & .LabelItem:last-child {
    margin-right: 0;
  }
`

const LabelGroup = ({children, ...rest}) => <StyledLabelGroup {...rest}>{transformChildren(children)}</StyledLabelGroup>

LabelGroup.defaultProps = {
  theme
}

LabelGroup.propTypes = {
  ...COMMON.propTypes
}

export default LabelGroup
