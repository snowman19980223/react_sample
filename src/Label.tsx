import styled, {css} from 'styled-components'
import {borderColor, BorderColorProps, variant} from 'styled-system'
import {COMMON, get, SystemCommonProps} from './constants'
import sx, {SxProp} from './sx'
import theme from './theme'
import {ComponentProps} from './utils/types'

const outlineStyles = css`
  margin-top: -1px; // offsets the 1px border
  margin-bottom: -1px; // offsets the 1px border
  color: ${get('colors.gray.6')};
  border: ${get('borderWidths.1')} solid ${get('colors.blackfade15')};
  box-shadow: none;
  ${borderColor};
  ${COMMON};
  background-color: transparent;
`

const sizeVariant = variant({
  variants: {
    small: {
      fontSize: 0,
      lineHeight: '16px',
      padding: '0px 8px'
    },
    medium: {
      fontSize: 0,
      lineHeight: '20px',
      padding: '0 8px'
    },
    large: {
      fontSize: 0,
      lineHeight: '24px',
      padding: '0 12px'
    },
    // corresponds to StateLabel fontSize/lineHeight/padding
    xl: {
      fontSize: 1,
      lineHeight: '16px',
      padding: '8px 12px'
    }
  }
})

const Label = styled.span<
  {
    variant?: 'small' | 'medium' | 'large' | 'xl'
    dropshadow?: boolean
    outline?: boolean
  } & BorderColorProps &
    SystemCommonProps &
    SxProp
>`
  display: inline-block;
  font-weight: ${get('fontWeights.semibold')};
  color: ${get('colors.white')};
  border-radius: ${get('radii.3')};

  &:hover {
    text-decoration: none;
  }

  ${sizeVariant}
  ${COMMON}
  ${props => (props.dropshadow ? 'box-shadow: inset 0 -1px 0 rgba(27, 31, 35, 0.12)' : '')}
  ${props => (props.outline ? outlineStyles : '')} // must be last to override other values
  ${sx}
`

Label.defaultProps = {
  theme,
  bg: 'gray.5',
  variant: 'medium'
}

export type LabelProps = ComponentProps<typeof Label>
export default Label
