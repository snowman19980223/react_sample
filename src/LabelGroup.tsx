import styled from 'styled-components'
import {COMMON, get, SystemCommonProps} from './constants'
import sx, {SxProp} from './sx'
import theme from './theme'
import {ComponentProps} from './utils/types'

const LabelGroup = styled.span<SystemCommonProps & SxProp>`
  ${COMMON}
  & * {
    margin-right: ${get('space.1')};
  }
  & *:last-child {
    margin-right: 0;
  }
  ${sx};
`

LabelGroup.defaultProps = {
  theme
}

export type LabelGroupProps = ComponentProps<typeof LabelGroup>
export default LabelGroup
