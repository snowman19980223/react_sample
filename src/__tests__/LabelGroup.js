import React from 'react'
import LabelGroup from '../LabelGroup'
import Label from '../Label'
import {render} from '../utils/testing'
import {COMMON} from '../constants'

const comp = (
  <LabelGroup>
    <Label>Default label</Label>
    <Label>Darker gray label</Label>
    <Label outline>Default outline label</Label>
  </LabelGroup>
)

describe('BranchName', () => {
  it('has default theme', () => {
    expect(LabelGroup).toSetDefaultTheme()
  })

  it('respects the "as" prop', () => {
    expect(render(<LabelGroup as="div" />).type).toEqual('div')
  })

  it('implements common system props', () => {
    expect(LabelGroup).toImplementSystemProps(COMMON)
  })

  it('matches snapshot', () => {
    expect(render(comp)).toMatchSnapshot()
  })
})
