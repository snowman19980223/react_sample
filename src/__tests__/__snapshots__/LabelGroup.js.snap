// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BranchName matches snapshot 1`] = `
.c0 .LabelItem {
  margin-right: 4px;
}

.c0 .LabelItem:last-child {
  margin-right: 0;
}

.c1 {
  display: inline-block;
  font-weight: 600;
  line-height: 1;
  color: #fff;
  border-radius: 2px;
  font-size: 12px;
  padding-left: 3px;
  padding-right: 3px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #6a737d;
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2 {
  display: inline-block;
  font-weight: 600;
  line-height: 1;
  color: #fff;
  border-radius: 2px;
  font-size: 12px;
  padding-left: 3px;
  padding-right: 3px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #6a737d;
  margin-top: -1px;
  margin-bottom: -1px;
  font-weight: 400;
  color: #586069;
  background-color: transparent;
  border: 1px solid rgba(27,31,35,0.15);
  box-shadow: none;
}

.c2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<span
  className="c0"
>
  <span
    className="c1 LabelItem"
  >
    Default label
  </span>
  <span
    className="c1 LabelItem"
  >
    Darker gray label
  </span>
  <span
    className="c2 LabelItem"
  >
    Default outline label
  </span>
</span>
`;
